lib/Target/Btor/BtorToBtorIRTranslation.cpp

```diff
@@ -0,0 +1,517 @@
+#include "Target/Btor/BtorToBtorIRTranslation.h"
+#include "Dialect/Btor/IR/BtorDialect.h"
+#include "Dialect/Btor/IR/BtorOps.h"
+
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/BuiltinOps.h"
+#include "mlir/IR/Dialect.h"
+#include "mlir/Translation.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/MemoryBuffer.h"
+#include "llvm/Support/SourceMgr.h"
+
+#include <iostream>
+#include <string>
+
+using namespace mlir;
+using namespace mlir::btor;
+
+void Deserialize::parseModelLine(Btor2Line *l) {
+  reachedLines[l->id] = l;
+  switch (l->tag) {
+  case BTOR2_TAG_bad:
+    bads.push_back(l);
+    break;
+
+  case BTOR2_TAG_constraint:
+    constraints.push_back(l);
+    break;
+
+  case BTOR2_TAG_init:
+    inits.push_back(l);
+    break;
+
+  case BTOR2_TAG_input:
+    inputs.push_back(l);
+    break;
+
+  case BTOR2_TAG_next:
+    nexts.push_back(l);
+    break;
+
+  case BTOR2_TAG_state:
+    states.push_back(l);
+    break;
+
+  default:
+    break;
+  }
+}
+
+bool Deserialize::parseModelIsSuccessful() {
+  if (!modelFile)
+    return false;
+  model = btor2parser_new();
+  if (!btor2parser_read_lines(model, modelFile)) {
+    std::cerr << "parse error at: " << btor2parser_error(model) << "\n";
+    return false;
+  }
+  // register each line that has been parsed
+  Btor2LineIterator it = btor2parser_iter_init(model);
+  Btor2Line *line;
+  while ((line = btor2parser_iter_next(&it))) {
+    parseModelLine(line);
+  }
+  // ensure each state has a next function
+  for (auto state : states) {
+    if (reachedLines.find(state->next) == reachedLines.end()) {
+      std::cerr << "state " << state->id << " without next function\n";
+      return false;
+    }
+  }
+
+  return true;
+}
+
+///===----------------------------------------------------------------------===//
+/// This function's goal is to create the MLIR Operation that corresponds to 
+/// the given Btor2Line*, cur, into the basic block designated by the provided 
+/// builder.
+///
+/// e.x:
+///     Operation * res = createMLIR(cur, cur->args);
+///
+///===----------------------------------------------------------------------===//
+Operation * Deserialize::createMLIR(const Btor2Line *line, const int64_t *kids) {
```

Comment:
somewhat annoying that `kids` is not a vector, or at the very least, `ArrayRef`

Reply:
also hard to add defensive assertions to check that the number of kids is as expected

Reply:
Will use an `ArrayRef` instead